### DISCUSSION TITLE
Fixed image urls that check security

### DISCUSSION THREAD

**Discussion by pauldambra** [supabase/supabase#12022] [Questions]
Would love people's help since I find myself learning too much at once and a bit stuck 😊 

I am building a site with next/vercel and supabase. Some users can upload photos associated with a particular user. Other users can view those photos if they have permission.

I'm using signed URLs but they're super slow. I'm seeing 1.5 seconds TTFB and can't cache because they change frequently. 

I don't want URLs that anonymous users can access. But I do want fixed URLs so I can cache etc.

Do I want a signed URL that doesn't expire or....?

**Comment by pauldambra** [comment]
So, my ideal world would be a bucket per "person"

```
-/personA
   -/imageOne
-/personB
   -/imageTwo
-/personC
  -/imageThree
```

then their URLs would be `${someprefix}.com/${personId}/${imageName}`

both person ID and imageName are known so my app can always create the URL from a template

each bucket/image is then protected so if `userOne` can only see images for `personC` then even if they guess the URLs they don't have access to persons A and B

**Comment by GaryAustin1** [comment]
If you have long lived signed URL you might as well have public urls with a random uuid in them.  Either way if the url is shared the image is accessible.  RLS protects all other aspects of the buckets/objects the same for public and private buckets so you can block operations like listing all files even on a public bucket.

**Reply by pauldambra** [reply]
Yep, I realised that the signed URLs is not a solution for me since posting. Thanks for confirming

Is there a way to have private image URLs on supabase?

**Reply by GaryAustin1** [reply]
You can't enforce RLS on any image URL method.  Once you have the URL, signed or public it is good for anyone who has it until it expires.

So you have signed URL, a public URL using a random UUID (hard to guess URL) and then to download the file data (which will need to meet RLS) and build your own image.

**Reply by pauldambra** [reply]
Wow TTFB on the download method is around 120ms compared to over 1000ms for a signed URL 🤯 

With a little image optimisation, I can live with downloading the image nad converting them to object URLs...

```
    useEffect(() => {
        const loadImage = async () => {
            const { data, error } = await supabase.storage
                .from('images')
                .download(props.image)

            if (data) {
                setImageURL(URL.createObjectURL(data))
            } else {
                setError('Error loading image: ' + JSON.stringify(error))
            }
        }
        if (!!props.image && !!props.student && !!supabase) {
            loadImage()
        }
    }, [props.image, props.student, supabase])
```

(for anyone that turns up here through Google, this is what I ended up doing) So that I can enforce RLS on access to the images

Thanks @GaryAustin1 💖
